### DISCUSSION TITLE
Messages being consumed in bursts

### DISCUSSION THREAD

**Discussion by brian7704** [pika/pika#1504] [Q&A]
I use Pika in my project [OpenTAKServer](https://github.com/brian7704/OpenTAKServer) and my consumer seems to be receiving messages in bursts. Ideally I'd like for it to consume new messages as fast as possible. However I'm not sure if the issue is with RabbitMQ, Pika, or both. I'm sure I've got something misconfigured. Here is the consumer.

https://github.com/brian7704/OpenTAKServer/blob/master/opentakserver/controllers/rabbitmq_client.py

https://github.com/brian7704/OpenTAKServer/blob/master/opentakserver/controllers/cot_controller.py

**Comment by lukebakken** [comment]
There's quite a bit of code to wade through. Things of note:
* You don't call the `basic_qos` method on your channel, which means RabbitMQ will send messages as fast as possible to your consumer, without limit. This is usually a bad idea, so I suggest reading about "QoS"/"Prefetch" for consumers and experimenting with different values.
* Pika is not thread-safe. This means that your `on_message` method will _block_ the thread running Pika's I/O loop until `on_message` returns. I don't know how long the work in that method takes, but if it's not extremely fast, all of the time, I suggest doing the work on a different thread, in combination with `basic_qos`, to limit the amount of work being done at any one time. See [this code](https://github.com/pika/pika/blob/main/examples/basic_consumer_threaded.py) - even though it uses `BlockingConnection`, the concepts are the same for `SelectConnection`

**Comment by brian7704** [comment]
Thanks for your reply and sorry for all the code. That consumer parses XML from client apps.

I tried playing around with different prefetch values but they didn't seem to have any effect. However I've now figured out that it takes anywhere from 1 - 4 seconds for my consumer to receive a message (this is before the consumer does any work) once it's been published. The consumer's work only takes about 1/10th of a second, so it's sitting idle while it waits for more messages. Then after about 4 seconds or so, all of the available messages in the queue (usually 3 or 4) all come through at once.

Could it have something to do with how I'm [publishing the messages](https://github.com/brian7704/OpenTAKServer/blob/28bd2d8679d6c748bbb3136421d7b2d3a2341e72/opentakserver/controllers/client_controller.py#L249)?

**Reply by lukebakken** [reply]
>  I've now figured out that it takes anywhere from 1 - 4 seconds for my consumer to receive a message once it's been published.

This could be due to:
* A very slow network
* A very slow machine running RabbitMQ
* Your code (most likely)

You could do the following to diagnose:
* Run Wireshark to capture when your consumer receives packets vs actually processes them. My guess is that the delay is in your consuming application - something blocks Pika's I/O loop and prevents it from processing incoming TCP data.
* Enable debug logging for Pika to see the same (https://github.com/pika/pika/blob/main/examples/basic_consumer_threaded.py#L11-L15)

If you could provide a project that I could run to see the same behavior, I could diagnose it quickly.

There is something that doesn't feel quite right about how you are creating a Thread for your Pika SelectConnection:

https://github.com/brian7704/OpenTAKServer/blob/master/opentakserver/controllers/rabbitmq_client.py#L20-L26

The instance members for `RabbitMQClient` will be created on the main program thread, not the thread running the `SelectConnection`'s I/O loop, correct? So you have cross-thread calls that aren't allowed in Pika.

`RabbitMQClient` should subclass `Thread`, and you should use [`add_callback_threadsafe`](https://github.com/pika/pika/blob/main/examples/basic_consumer_threaded.py#L30-L37) any time you need to schedule work to be done on the thread running Pika.

I realize this is a PITA but Pika is low-level and not thread-safe at this time.

**Reply by brian7704** [reply]
I finally solved this issue. It seemed to be caused by the Python GIL. By breaking out my producer and consumer into separate processes instead of separate threads, messages are consumed as expected.
